### ISSUE TITLE
Regression: Poetry uses the wrong version of Python

### ISSUE THREAD

**Issue by sargunv** [Homebrew/homebrew-core#120502]
### ~~`brew gist-logs <formula>` link OR~~ `brew config` AND `brew doctor` output

```shell
HOMEBREW_VERSION: 3.6.18-42-g4abc51f
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4abc51f32869511eaa29c7feaa30312aec70e1d2
Last commit: 15 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b52e0ef46820f2454d0d2fe5b9c82af8038ecc53
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.0 => /opt/homebrew/bin/git
Curl: 7.84.0 => /usr/bin/curl
macOS: 13.0.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
Rosetta 2: false
```

```txt
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Trying to `poetry install` with Python v3.8.13 in my PATH, in a project that requires Python >=3.8.5,<3.10.0 according to its pyproject.toml.

The python in my PATH is v3.8.13 (from pyenv):

<img width="327" alt="CleanShot 2023-01-12 at 17 18 21@2x" src="https://user-images.githubusercontent.com/1320357/212214912-d09e9dc5-34d9-4a2b-abb4-600d0e8bec1a.png">

The python required by my project using Poetry is >=3.8.5,<3.10.0

<img width="572" alt="CleanShot 2023-01-12 at 17 22 18@2x" src="https://user-images.githubusercontent.com/1320357/212215354-46f5be8e-f28e-41ac-9caa-1878692972a1.png">

So, a `poetry install` in this project should work. 

### What happened (include all command output)?

Poetry somehow runs with Python 3.11.1, so `poetry install` fails as 3.11.1 does not meet the constraints in the project:

<img width="909" alt="CleanShot 2023-01-12 at 17 23 25@2x" src="https://user-images.githubusercontent.com/1320357/212215565-e9e5ddd6-f163-4cc8-a9b6-ff1cb1c7119e.png">

Even `poetry env list` shows the venv is _named_ as if it's a 3.8 venv:

<img width="293" alt="CleanShot 2023-01-12 at 17 25 05@2x" src="https://user-images.githubusercontent.com/1320357/212215807-34bb06e1-cb61-4fdb-9cc9-f1dc807b7bd4.png">

With `poetry env info` we can see that poetry is ignoring my PATH and instead using the python from its homebrew package:

<img width="502" alt="CleanShot 2023-01-12 at 17 26 10@2x" src="https://user-images.githubusercontent.com/1320357/212215878-293e8dd5-d0f7-4124-8a36-ac88c0779917.png">

~~And here's the culprit. The `#!` of the poetry bin hardcodes the Python path.~~ EDIT: official installer does too, and does not exhibit the same bug.

<img width="502" alt="CleanShot 2023-01-12 at 17 29 33@2x" src="https://user-images.githubusercontent.com/1320357/212216275-c1b9a30f-0088-44f5-85ad-96e197ee387f.png">

### What did you expect to happen?

I expected Poetry to use the Python from my PATH to install the project I'm working on (that's compatible with the Python from my PATH).

### Step-by-step reproduction instructions (by running `brew` commands)

```
brew install pyenv
pyenv install 3.8.13
pyenv global 3.8.13
brew install poetry
poetry init -n --name test --python 3.8.13
poetry install
```

**Comment by sargunv** [comment]
This is a regression of https://github.com/Homebrew/homebrew-core/issues/62910, which was fixed in https://github.com/Homebrew/homebrew-core/commit/c9b23e825ff65e520ca706e77cc11db577bbcd7a. I believe it was broken in https://github.com/Homebrew/homebrew-core/commit/763ca9d967c21667064396fa7b2988526f2a1e96.

**Comment by sargunv** [comment]
Workaround: `poetry env remove --all` and `poetry env use "$(which python3)"` to get the right env.

<img width="763" alt="CleanShot 2023-01-12 at 17 37 47@2x" src="https://user-images.githubusercontent.com/1320357/212217193-b3856089-a7d5-4cda-b802-7c364bc4f353.png">

**Comment by sargunv** [comment]
Interestingly, in [the commit](https://github.com/Homebrew/homebrew-core/commit/763ca9d967c21667064396fa7b2988526f2a1e96) that I believe broke it, there's a test that attempts to test for #62910, but the test uses `poetry env use` to explicitly select a Python path and therefore doesn't trigger the bug

https://github.com/Homebrew/homebrew-core/blob/763ca9d967c21667064396fa7b2988526f2a1e96/Formula/poetry.rb#L205-L214

**Comment by cho-m** [comment]
Homebrew's installation is similar to the documented manual installation (https://python-poetry.org/docs/#installing-manually) as we essentially do a `pip install` inside a venv.

I think the official installer also just installs into a venv and adds the corresponding venv shebang.

Given above, this seems more like an upstream issue, maybe https://github.com/python-poetry/poetry/issues/7158

**Comment by sargunv** [comment]
This may be fixable upstream, but I'm unable to replicate the problem when poetry is installed with the official install instructions. _Something_ is different between what Brew is doing and what the official installer is doing.

However, you are correct that the one from the official installer also hardcodes the python in the shebang:

![image](https://user-images.githubusercontent.com/1320357/212244426-5494b29c-94cf-4121-875d-52a689cdde8e.png)

**Comment by sargunv** [comment]
Will keep an eye on https://github.com/python-poetry/poetry/pull/7221 to see if the upstream fix also applies here though

**Comment by the-serious-programmer** [comment]
@sargunv thanks so much for the mentioned workaround:
```
Workaround: poetry env remove --all and poetry env use "$(which python3)" to get the right env.
```

I was literally spending hours and hours on why my poetry kept falling back to python 3.11 instead of 3.9. 
Originally came from: https://github.com/python-poetry/poetry/issues/7158

**Comment by sargunv** [comment]
should be fixed in 1.4.0

**Comment by rtviii** [comment]
Still having this issue.
